### PR TITLE
Introduce CheckpointManager

### PR DESCRIPTION
**Pull request by cbalioglu**
This PR introduces `CheckpointManager`, and performs some nit docstr updates and variable renaming.

### CODE REVIEW

File: src/fairseq2/checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+                raise RuntimeError(
+                    f"The checkpoint directory for step {step_nr} cannot be created. See inner exception for details."
+                ) from ex
+
+        self.gang.barrier()
+
+        checkpoint_file = step_dir.joinpath(f"rank_{self.gang.rank}.pt")
+
+        try:
+            torch.save(checkpoint, checkpoint_file)
```
Comment (shagunsodhani) [inline src/fairseq2/checkpoint.py:137]: For DDP, wouldn't we end up saving redundant copies of the model ?

File: src/fairseq2/checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+
+    @finaloverride
+    def save_checkpoint(self, step_nr: int, state: Dict[str, Any]) -> None:
+        checkpoint = {"step_nr": step_nr, "data": state}
+
+        step_dir = self.checkpoint_dir.joinpath(f"step_{step_nr}")
+
+        if self.gang.rank == 0:
+            try:
+                step_dir.mkdir(parents=True, exist_ok=True)
```
Comment (shagunsodhani) [inline src/fairseq2/checkpoint.py:126]: When checkpointing FSDP models on non-nfs / local partitions, we may want to execute this step on all the nodes.

File: src/fairseq2/checkpoint.py (hunk 1 above)
Comment (cbalioglu) [reply src/fairseq2/checkpoint.py:137]: For `CheckpointManager` the provided `state` is an opaque dict. It does not really care what it contains. In a DDP setting, the trainer would do something like:

```
state_dict = {...}

if rank == 0:
    state_dict["model"] = model.state_dict()

checkpoint_manager.save_checkpoint(1234, state_dict)
```

File: src/fairseq2/checkpoint.py (hunk 2 above)
Comment (cbalioglu) [reply src/fairseq2/checkpoint.py:126]: What do you mean by "non-nfs/local" partitions?

File: src/fairseq2/checkpoint.py (hunk 1 above)
Comment (shagunsodhani) [reply src/fairseq2/checkpoint.py:137]: Okay so it is the responsibility of the caller to make sure they do not call `checkpoint_manager.save_checkpoint(1234, state_dict)` across multiple ranks when training with ddp.

File: src/fairseq2/checkpoint.py (hunk 2 above)
Comment (shagunsodhani) [reply src/fairseq2/checkpoint.py:126]: I was thinking of the case where we are checkpointing shards when using FSDP. In that case, all ranks will checkpoint their shard to the disk. We are creating the directory only in rank=0 process. Now if model was saved to a partition that is local to each node (i.e. physical machine), the checkpoint directory would only exist for the node with rank=0 process.

Comment (cbalioglu) [reply src/fairseq2/checkpoint.py:126]: Oh I see what you mean. In majority of cases, nodes share a common distributed file system (e.g NFS, fsx), but I think it is a fair ask to support local disk setup. Let me revise this implementation to address your feedback. Thanks!

**Pull request merged**